packages/redis/test/Adaptor.test.js

```diff
@@ -0,0 +1,334 @@
+    } catch (error) {
+      expect(error.message).to.eql('TypeError: Invalid argument type');
+    }
+  });
+  it('should throw an error if key is not specified', async () => {
+    setMockClient({
+      set: async () => {
+        throw new Error('TypeError: Invalid argument type');
+      },
+    });
+
+    const state = {};
+    try {
+      await set('', 'kwaku')(state);
+    } catch (error) {
+      expect(error.message).to.eql('TypeError: Invalid argument type');
+    }
+  });
+});
+
+describe('hset', () => {
+  it('should set fields value for a specified key', async () => {
+    setMockClient({
+      hSet: async () => 1,
+    });
+
+    const state = {};
+    const result = await hset('patient', { name: 'kwaku' })(state);
+    expect(result.data).to.eql(1);
+  });
+  it('should expand references', async () => {
+    setMockClient({
+      hSet: async () => 1,
+    });
+    const state = { key: 'patient', value: { name: 'mtuchi' } };
+    const result = await hset(
+      s => s.key,
+      s => s.value
+    )(state);
+    expect(result.data).to.eql(1);
+  });
+  it('should throw an error if key and value is not specified', async () => {
+    setMockClient({
+      hSet: async () => {
+        throw new Error(
+          'TypeError: Cannot convert undefined or null to object'
+        );
+      },
+    });
+
+    const state = {};
+    try {
+      await hset()(state);
+    } catch (error) {
+      expect(error.message).to.eql(
+        'TypeError: Cannot convert undefined or null to object'
+      );
+    }
+  });
+  it('should throw an error if value is not specified', async () => {
+    setMockClient({
+      hSet: async () => {
+        throw new Error(
+          'TypeError: Cannot convert undefined or null to object'
+        );
+      },
+    });
+
+    const state = {};
+    try {
+      await hset('patient')(state);
+    } catch (error) {
+      expect(error.message).to.eql(
+        'TypeError: Cannot convert undefined or null to object'
+      );
+    }
+  });
+  it('should set an empt key if key is not specified', async () => {
+    setMockClient({
+      hSet: async () => 1,
+    });
+
+    const state = {};
+    const result = await hset('', { name: 'kwaku' })(state);
+    expect(result.data).to.eql(1);
+  });
+});
+describe('scan', () => {
+  it('should return empty array if key not found', async () => {
+    setMockClient({
+      scan: async () => ({ keys: [], cursor: 0 }),
+    });
+    const state = {};
+    const result = await scan('*:abc*')(state);
+    expect(result.data).to.eql([]);
+    expect(result.cursor).to.eql(0);
+  });
+  it('should return collection of keys', async () => {
+    setMockClient({
+      scan: async () => ({ keys: ['noderedis:animals:1'], cursor: 0 }),
+    });
+
+    const state = {};
+    const result = await scan('*:animals*')(state);
+    expect(result.data).to.eql(['noderedis:animals:1']);
+    expect(result.cursor).to.eql(0);
+  });
+  it('should return all hash keys', async () => {
```

Comment:
>Is this test exercising adaptor code sufficiently?

Can you shed more light ?, i am not sure if understand the question

Reply:
Yes - are you familiar with the idea of test coverage?

When you run a function, Javascript will execute statement after statement. Depending on the inputs to the function, the execution will take a different path through the code. Maybe it'll enter this if block, maybe it;'ll go around this loop 20 times, maybe it'll throw this exception.

The idea of a _suite_  of unit tests is that, between all the tests, every possible path of  execution is followed.

Now I'm not asking for THAT. But I _am_ asking you to think carefully about whether your suite of tests is triggering the right paths of execution through the function. And also whether your tests are correctly validating that those paths are followed.

I asked the question because these tests look very similar, and all they really do is return different values out of the mock function. I'm just wondering if we should be doing more.

For example: looking closely at this test and the underlying code, I think really that this is a test that `type` defaults to `hash` (so maybe the test name is `should return hash keys by default`). To test that properly, you need to work out whether the adaptor has set the `TYPE` option to `hash`.

Just like with our salesforce PR recently, the questions are:
* What are you ACTUALLY testing?
* How will you verify that the thing you wanted to happen actually happened?

Reply:
I think the fix here is to name the test `should return hash keys by default` and to assert that options.TYPE is set. That then is a very useful test and I am happy.